osu.Game/Scoring/ScoreInfo.cs

```diff
@@ -36,27 +35,49 @@ public class ScoreInfo : IHasFiles<ScoreFileInfo>, IHasPrimaryKey, ISoftDelete
 
         public int RulesetID { get; set; }
 
-        public RulesetInfo Ruleset { get; set; }
+        public virtual RulesetInfo Ruleset { get; set; }
 
-        [NotMapped, JsonIgnore]
+        private Mod[] mods;
+
+        [NotMapped]
         public Mod[] Mods
         {
             get
             {
-                if (ModsString == null)
+                if (mods != null) return mods;
+
+                if (modsJson == null)
                     return Array.Empty<Mod>();
 
-                var deserialized = JsonConvert.DeserializeObject<string[]>(ModsString);
-                return Ruleset.CreateInstance().GetAllMods().Where(mod => deserialized.Any(d => d == mod.ShortenedName)).ToArray();
+                return getModsFromRuleset(JsonConvert.DeserializeObject<DeserializedMod[]>(modsJson));
+            }
+            set
+            {
+                modsJson = null;
+                mods = value;
             }
-            set => ModsString = JsonConvert.SerializeObject(value.Select(m => m.ShortenedName).ToArray());
         }
 
-        public string ModsString { get; set; }
+        private Mod[] getModsFromRuleset(DeserializedMod[] mods) => Ruleset.CreateInstance().GetAllMods().Where(mod => mods.Any(d => d.Acronym == mod.Acronym)).ToArray();
+
+        private string modsJson;
+
+        [Column("Mods")]
+        public string ModsJson
+        {
+            get => modsJson ?? (modsJson = JsonConvert.SerializeObject(mods));
+            set
+            {
+                modsJson = value;
+
+                // we potentially can't update this yet due to Ruleset being late-bound, so instead update on read as necessary.
+                mods = null;
+            }
+        }
 
-        [NotMapped, JsonIgnore]
         public User User;
 
+        [Column("User")]
         public string UserString
```

Comment:
This should be named `UserJson` to be consistent with everything else in this file. Requires new DB migration

Reply:
It's a string though..

Reply:
Yeah, so is `StatisticsJson` and `ModsJson`

Reply:
Oh... I misunderstood